src/arrays/arraymath.jl

```diff
@@ -42,69 +46,96 @@ end
-        return QuArray(coeffs(qarr1)+coeffs(qarr2), bases(qarr1))
+        sc = coeffs(qarr1) + coeffs(qarr2)
+        QAT = similar_type(promote_type(typeof(qarr1), typeof(qarr2))) 
+        return QAT(sc, bases(qarr1))
     else
         error("Bases not compatible")
     end
 end
 
-function -(qarr1::AbstractQuArray, qarr2::AbstractQuArray)
+function -{B<:AbstractBasis}(qarr1::AbstractQuArray{B}, qarr2::AbstractQuArray{B})
     if bases(qarr1) == bases(qarr2)
-        return QuArray(coeffs(qarr1)-coeffs(qarr2), bases(qarr1))
+        sc = coeffs(qarr1) - coeffs(qarr2)
+        QAT = similar_type(promote_type(typeof(qarr1), typeof(qarr2))) 
+        return QAT(sc, bases(qarr1))
     else
         error("Bases not compatible")
     end
 end
 
++(ct1::CTranspose, ct2::CTranspose) = (ct1.qarr+ct2.qarr)'
+-(ct1::CTranspose, ct2::CTranspose) = (ct1.qarr-ct2.qarr)'
+
+
 # scaling
-Base.scale!(num::Number, qarr::QuArray) = (scale!(num, rawcoeffs(qarr)); return qarr)
+Base.scale!(num::Number, qarr::AbstractQuArray) = (scale!(num, rawcoeffs(qarr)); return qarr)
 Base.scale!(num::Number, ct::CTranspose) = CTranspose(scale!(num', ct.qarr))
-Base.scale!(qarr::Union(QuArray,CTranspose), num::Number) = scale!(num, qarr)
+Base.scale!(qarr::AbstractQuArray, num::Number) = scale!(num, qarr)
 
-Base.scale(num::Number, qarr::QuArray) = QuArray(scale(num, rawcoeffs(qarr)), rawbases(qarr))
+function Base.scale(num::Number, qarr::AbstractQuArray)
+    fc = scale(num, rawcoeffs(qarr))
+    QAT = QuBase.similar_type(typeof(qarr))
+    return QAT(fc, bases(qarr))
+end
 Base.scale(num::Number, ct::CTranspose) = CTranspose(scale(num', ct.qarr))
-Base.scale(qarr::Union(QuArray,CTranspose), num::Number) = scale(num, qarr)
+Base.scale(qarr::AbstractQuArray, num::Number) = scale(num, qarr)
 
-*(num::Number, qarr::Union(QuArray,CTranspose)) = scale(num, qarr)
-*(qarr::Union(QuArray,CTranspose), num::Number) = scale(qarr, num)
-/(qarr::Union(QuArray,CTranspose), num::Number) = scale(1/num, qarr)
+*(num::Number, qarr::AbstractQuArray) = scale(num, qarr)
+*(qarr::AbstractQuArray, num::Number) = scale(qarr, num)
+/(qarr::AbstractQuArray, num::Number) = scale(1/num, qarr)
 
-# sparse to dense
-Base.full(qarr::AbstractQuMatrix) = QuArray(full(coeffs(qarr)),bases(qarr))
-
-# exponential of dense matrix
-Base.expm(qarr::AbstractQuMatrix) = QuArray(expm(full(coeffs(qarr))),bases(qarr))
 
+# matrix operations returning a scalar
 # normalization
-Base.norm(qarr::Union(QuArray,CTranspose)) = vecnorm(rawcoeffs(qarr))
+Base.norm(qarr::AbstractQuArray) = vecnorm(rawcoeffs(qarr))
 
-function normalize!(qarr::Union(QuArray,CTranspose))
+function normalize!(qarr::AbstractQuArray)
     scale!(1/norm(qarr), rawcoeffs(qarr))
     return qarr
 end
 
-normalize(qarr::Union(QuArray,CTranspose)) = normalize!(copy(qarr))
+normalize(qarr::AbstractQuArray) = normalize!(copy(qarr))
+
+
+# matrix operations returning an array
+# sparse to dense
+function Base.full(qarr::AbstractQuMatrix)
```

Comment:
So like `@applycoeffs f(qa)` --> `QuArray(f(coeffs(qa)), bases(qa))`? That could work. OTOH, making users write out `QuArray(f(coeffs(qa), bases(qa))` might result in more "deliberate" code, as the intent about the return type of `f` is clearer. I could go either way, honestly.